File: sdk/lib/csharp/ConstraintArithmetic/ConstraintArithmetic.cs
using System;
using System.Collections.Generic;
using static Solver;

#pragma warning disable CS0660, CS0661

namespace ConstraintArithmetic {
    public struct ArithmeticConstraint {
        public float LeftConstant;
        public float[] LeftCoefficients;

        public float RightConstant;
        public float[] RightCoefficients;

        public ArithmeticConstraint(float leftConstant, float[] leftCoefficients, float rightConstant, float[] rightCoefficients) {
            LeftConstant = leftConstant;
            LeftCoefficients = leftCoefficients;

            RightConstant = rightConstant;
            RightCoefficients = rightCoefficients;
        }
    }

    public enum ArithmeticInequality {
        Equal,
        LessThanOrEqual,
        GreaterThanOrEqual
    }

    public class ArithmeticContext {
        public List<bool> Variables = new List<bool>();

        public List<ArithmeticConstraint> Constraints = new List<ArithmeticConstraint>();

        public ArithmeticExpression Objective;

        public int[] SolutionVariableIndices;
        public float[] SolutionConstants;

        public static ArithmeticConstraint CreateConstraint(ArithmeticExpression left, ArithmeticInequality inequality, ArithmeticExpression right) {
            var context = left.Context;

            switch(inequality) {
                case ArithmeticInequality.Equal: break;

                case ArithmeticInequality.LessThanOrEqual: {
                    var slackVariable = context.Variables.Count;

                    context.Variables.Add(false);

                    Array.Resize(ref left.Coefficients, context.Variables.Count);

                    left.Coefficients[slackVariable] = 1.0f;
                } break;

                case ArithmeticInequality.GreaterThanOrEqual: {
                    var slackVariable = context.Variables.Count;

                    context.Variables.Add(false);

                    Array.Resize(ref left.Coefficients, context.Variables.Count);

                    left.Coefficients[slackVariable] = -1.0f;
                } break;
            }

            return new ArithmeticConstraint(left.Constant, left.Coefficients, right.Constant, right.Coefficients);
        }

        public void AddConstraints(params ArithmeticConstraint[] constraints) {
            Constraints.AddRange(constraints);
        }

        public void Solve() {
            // Create full-sized tableau
            var objectiveConstant = Objective.Constant;
            var objectiveCoefficients = new float[Variables.Count];

            if(Objective.Coefficients != null) {
                for(var i = 0; i < Objective.Coefficients.Length; i += 1) {
                    objectiveCoefficients[i] = Objective.Coefficients[i];
                }
            }

            var constraintConstants = new float[Constraints.Count];

            var constraintCoefficients = new float[Constraints.Count * Variables.Count];

            for(var i = 0; i < Constraints.Count; i += 1) {
                var constraint = Constraints[i];

                constraintConstants[i] = constraint.LeftConstant;
                constraintConstants[i] -= constraint.RightConstant;

                if(constraint.LeftCoefficients != null) {
                    for(var j = 0; j < constraint.LeftCoefficients.Length; j += 1) {
                        constraintCoefficients[i * Variables.Count + j] = constraint.LeftCoefficients[j];
                    }
                }

                if(constraint.RightCoefficients != null) {
                    for(var j = 0; j < constraint.RightCoefficients.Length; j += 1) {
                        constraintCoefficients[i * Variables.Count + j] -= constraint.RightCoefficients[j];
                    }
                }
            }

            var constraintVariableIndices = new UIntPtr[Constraints.Count];

            // Run solver
            solve(
                (UIntPtr)Variables.Count,
                (UIntPtr)Constraints.Count,
                Variables.ToArray(),
                ref objectiveConstant,
                objectiveCoefficients,
                constraintVariableIndices,
                constraintConstants,
                constraintCoefficients
            );

            // Extract variable values from solution
            SolutionVariableIndices = new int[Constraints.Count];

            for(var i = 0; i < Constraints.Count; i += 1) {
                SolutionVariableIndices[i] = (int)constraintVariableIndices[i];
            }

            SolutionConstants = constraintConstants;
        }

        public void Minimize(ArithmeticExpression expression) {
            Objective = Objective + expression;
        }

        public void Maximize(ArithmeticExpression expression) {
            Objective = Objective - expression;
        }

        public void Minimize(ArithmeticTerm term) {
            Minimize(term);
        }

        public void Maximize(ArithmeticTerm term) {
            Maximize(term);
        }
    }

    public struct ArithmeticExpression {
        public ArithmeticContext Context;

        public float Constant;

        public float[] Coefficients;

        public ArithmeticExpression(ArithmeticContext context, float constant) {
            Context = context;
            Constant = constant;
            Coefficients = null;
        }

        public ArithmeticExpression(ArithmeticContext context, float[] coefficients) {
            Context = context;
            Constant = 0.0f;
            Coefficients = coefficients;
        }

        public ArithmeticExpression(ArithmeticContext context, float constant, float[] coefficients) {
            Context = context;
            Constant = constant;
            Coefficients = coefficients;
        }

        // Expression-Expression operators

        public static ArithmeticExpression operator +(ArithmeticExpression a, ArithmeticExpression b) {
            var Context = a.Context;

            float[] coefficients;
            if(a.Coefficients.Length > b.Coefficients.Length) {
                for(var i = 0; i < a.Coefficients.Length; i += 1) {
                    a.Coefficients[i] += b.Coefficients[i];
                }

                coefficients = a.Coefficients;
            } else {
                for(var i = 0; i < b.Coefficients.Length; i += 1) {
                    b.Coefficients[i] += a.Coefficients[i];
                }

                coefficients = b.Coefficients;
            }

            return new ArithmeticExpression(Context, a.Constant + b.Constant, coefficients);
        }

        public static ArithmeticExpression operator -(ArithmeticExpression expression) {
            expression.Constant = -expression.Constant;

            for(var i = 0; i < expression.Coefficients.Length; i += 1) {
                expression.Coefficients[i] = -expression.Coefficients[i];
            }

            return expression;
        }

        public static ArithmeticExpression operator -(ArithmeticExpression a, ArithmeticExpression b) {
            return a + -b;
        }

        public static ArithmeticConstraint operator ==(ArithmeticExpression left, ArithmeticExpression right) {
            return ArithmeticContext.CreateConstraint(left, ArithmeticInequality.Equal, right);
        }

        public static ArithmeticConstraint operator !=(ArithmeticExpression left, ArithmeticExpression right) {
            throw new NotSupportedException();
        }

        public static ArithmeticConstraint operator <=(ArithmeticExpression left, ArithmeticExpression right) {
            return ArithmeticContext.CreateConstraint(left, ArithmeticInequality.LessThanOrEqual, right);
        }

        public static ArithmeticConstraint operator >=(ArithmeticExpression left, ArithmeticExpression right) {
            return ArithmeticContext.CreateConstraint(left, ArithmeticInequality.GreaterThanOrEqual, right);
        }

        // Expression-float operators

        public static ArithmeticExpression operator +(ArithmeticExpression expression, float constant) {
            expression.Constant += constant;

            return expression;
        }

        public static ArithmeticExpression operator +(float constant, ArithmeticExpression expression) {
            return expression + constant;
        }

        public static ArithmeticExpression operator -(ArithmeticExpression expression, float constant) {
            return expression + -constant;
        }

        public static ArithmeticExpression operator -(float constant, ArithmeticExpression expression) {
            return expression + -constant;
        }

        public static ArithmeticExpression operator *(ArithmeticExpression expression, float constant) {
            expression.Constant *= constant;

            for(var i = 0; i < expression.Coefficients.Length; i += 1) {
                expression.Coefficients[i] *= constant;
            }

            return expression;
        }

        public static ArithmeticExpression operator *(float constant, ArithmeticExpression expression) {
            return expression * constant;
        }

        public static ArithmeticExpression operator /(ArithmeticExpression expression, float constant) {
            return expression * (1.0f / constant);
        }

        public static ArithmeticConstraint operator ==(ArithmeticExpression left, float right) {
            var rightExpression = new ArithmeticExpression(left.Context, right);

            return left == rightExpression;
        }

        public static ArithmeticConstraint operator ==(float left, ArithmeticExpression right) {
            var leftExpression = new ArithmeticExpression(right.Context, left);

            return leftExpression == right;
        }

        public static ArithmeticConstraint operator !=(ArithmeticExpression left, float right) {
            throw new NotSupportedException();
        }

        public static ArithmeticConstraint operator !=(float left, ArithmeticExpression right) {
            throw new NotSupportedException();
        }

        public static ArithmeticConstraint operator <=(ArithmeticExpression left, float right) {
            var rightExpression = new ArithmeticExpression(left.Context, right);

            return left <= rightExpression;
        }

        public static ArithmeticConstraint operator <=(float left, ArithmeticExpression right) {
            return right >= left;
        }

        public static ArithmeticConstraint operator >=(ArithmeticExpression left, float right) {
            var rightExpression = new ArithmeticExpression(left.Context, right);

            return left >= rightExpression;
        }

        public static ArithmeticConstraint operator >=(float left, ArithmeticExpression right) {
            return right <= left;
        }
    }

    public struct ArithmeticTerm {
        public ArithmeticContext Context;

        public int VariableIndex;

        public float Coefficient;

        public ArithmeticTerm(ArithmeticContext context, int variableIndex, float coefficient) {
            Context = context;
            VariableIndex = variableIndex;
            Coefficient = coefficient;
        }

        // Term-Term operators

        public static ArithmeticExpression operator +(ArithmeticTerm a, ArithmeticTerm b) {
            var context = a.Context;

            var coefficients = new float[context.Variables.Count];

            coefficients[a.VariableIndex] = a.Coefficient;
            coefficients[b.VariableIndex] += b.Coefficient;

            return new ArithmeticExpression(context, coefficients);
        }

        public static ArithmeticTerm operator-(ArithmeticTerm term) {
            term.Coefficient = -term.Coefficient;

            return term;
        }

        public static ArithmeticExpression operator -(ArithmeticTerm a, ArithmeticTerm b) {
            return a + -b;
        }

        public static ArithmeticConstraint operator ==(ArithmeticTerm left, ArithmeticTerm right) {
            var leftExpression = (ArithmeticExpression)left;

            var rightExpression = (ArithmeticExpression)right;

            return leftExpression == rightExpression;
        }

        public static ArithmeticConstraint operator !=(ArithmeticTerm left, ArithmeticTerm right) {
            throw new NotSupportedException();
        }

        public static ArithmeticConstraint operator <=(ArithmeticTerm left, ArithmeticTerm right) {
            var leftExpression = (ArithmeticExpression)left;

            var rightExpression = (ArithmeticExpression)right;

            return leftExpression <= rightExpression;
        }

        public static ArithmeticConstraint operator >=(ArithmeticTerm left, ArithmeticTerm right) {
            var leftExpression = (ArithmeticExpression)left;

            var rightExpression = (ArithmeticExpression)right;

            return leftExpression >= rightExpression;
        }

        // Term-float operators

        public static ArithmeticExpression operator +(ArithmeticTerm term, float constant) {
            var context = term.Context;

            var coefficients = new float[context.Variables.Count];

            coefficients[term.VariableIndex] = term.Coefficient;

            return new ArithmeticExpression(context, constant, coefficients);
        }

        public static ArithmeticExpression operator +(float constant, ArithmeticTerm term) {
            return term + constant;
        }

        public static ArithmeticExpression operator -(ArithmeticTerm term, float constant) {
            return term + -constant;
        }

        public static ArithmeticExpression operator -(float constant, ArithmeticTerm term) {
            return term + -constant;
        }

        public static ArithmeticTerm operator *(ArithmeticTerm term, float constant) {
            term.Coefficient *= constant;

            return term;
        }

        public static ArithmeticTerm operator *(float constant, ArithmeticTerm term) {
            return term * constant;
        }

        public static ArithmeticTerm operator /(ArithmeticTerm term, float constant) {
            return term * (1.0f / constant);
        }

        // Type Conversions

        public static implicit operator ArithmeticExpression(ArithmeticTerm term) {
            var coefficients = new float[term.Context.Variables.Count];

            coefficients[term.VariableIndex] = term.Coefficient;

            return new ArithmeticExpression(term.Context, coefficients);
        }
    }

    public struct ArithmeticVariable {
        public ArithmeticContext Context;

        public int Index;

        public float Value {
            get {
                for(var i = 0; i < Context.Variables.Count; i += 1) {
                    if(Context.SolutionVariableIndices[i] == Index) {
                        return Context.SolutionConstants[i];
                    }
                }

                return 0.0f;
            }
        }

        public ArithmeticVariable(ArithmeticContext context) {
            Context = context;

            Index = Context.Variables.Count;

            Context.Variables.Add(true);
        }

        // Variable-Variable operators

        public static ArithmeticExpression operator+(ArithmeticVariable a, ArithmeticVariable b) {
            var context = a.Context;

            var coefficients = new float[context.Variables.Count];

            coefficients[a.Index] = 1.0f;
            coefficients[b.Index] += 1.0f;

            return new ArithmeticExpression(context, coefficients);
        }

        public static ArithmeticTerm operator-(ArithmeticVariable variable) {
            return new ArithmeticTerm(variable.Context, variable.Index, -1.0f);
        }

        public static ArithmeticExpression operator-(ArithmeticVariable a, ArithmeticVariable b) {
            return a + -b;
        }

        public static ArithmeticConstraint operator==(ArithmeticVariable left, ArithmeticVariable right) {
            var leftExpression = (ArithmeticExpression)left;

            var rightExpression = (ArithmeticExpression)right;

            return leftExpression == rightExpression;
        }

        public static ArithmeticConstraint operator!=(ArithmeticVariable left, ArithmeticVariable right) {
            throw new NotSupportedException();
        }

        public static ArithmeticConstraint operator<=(ArithmeticVariable left, ArithmeticVariable right) {
            var leftExpression = (ArithmeticExpression)left;

            var rightExpression = (ArithmeticExpression)right;

            return leftExpression <= rightExpression;
        }

        public static ArithmeticConstraint operator>=(ArithmeticVariable left, ArithmeticVariable right) {
            return right <= left;
        }

        // Variable-float operators

        public static ArithmeticExpression operator+(ArithmeticVariable variable, float constant) {
            var coefficients = new float[variable.Context.Variables.Count];

            coefficients[variable.Index] = 1.0f;

            return new ArithmeticExpression(variable.Context, constant, coefficients);
        }

        public static ArithmeticExpression operator+(float constant, ArithmeticVariable variable) {
            return variable + constant;
        }

        public static ArithmeticExpression operator-(ArithmeticVariable variable, float constant) {
            return variable + -constant;
        }

        public static ArithmeticExpression operator-(float constant, ArithmeticVariable variable) {
            return constant + -variable;
        }

        public static ArithmeticTerm operator*(ArithmeticVariable variable, float coefficient) {
            return new ArithmeticTerm(variable.Context, variable.Index, coefficient);
        }

        public static ArithmeticTerm operator*(float coefficient, ArithmeticVariable variable) {
            return variable * coefficient;
        }

        public static ArithmeticTerm operator/(ArithmeticVariable variable, float coefficient) {
            return variable * (1.0f / coefficient);
        }

        public static ArithmeticConstraint operator==(ArithmeticVariable left, float right) {
            var leftExpression = (ArithmeticExpression)left;

            var rightExpression = new ArithmeticExpression(left.Context, right);

            return leftExpression == rightExpression;
        }

        public static ArithmeticConstraint operator==(float left, ArithmeticVariable right) {
            return right == left;
        }

        public static ArithmeticConstraint operator!=(ArithmeticVariable left, float right) {
            throw new NotSupportedException();
        }

        public static ArithmeticConstraint operator!=(float left, ArithmeticVariable right) {
            throw new NotSupportedException();
        }

        public static ArithmeticConstraint operator<=(ArithmeticVariable left, float right) {
            var leftExpression = (ArithmeticExpression)left;

            var rightExpression = new ArithmeticExpression(left.Context, right);

            return leftExpression <= rightExpression;
        }

        public static ArithmeticConstraint operator<=(float left, ArithmeticVariable right) {    
            return right >= left;
        }

        public static ArithmeticConstraint operator>=(ArithmeticVariable left, float right) {
            var leftExpression = (ArithmeticExpression)left;

            var rightExpression = new ArithmeticExpression(left.Context, right);

            return leftExpression >= rightExpression;
        }

        public static ArithmeticConstraint operator>=(float left, ArithmeticVariable right) {
            return right <= left;
        }

        // Type conversions

        public static implicit operator ArithmeticTerm(ArithmeticVariable variable) {
            return new ArithmeticTerm(variable.Context, variable.Index, 1.0f);
        }

        public static implicit operator ArithmeticExpression(ArithmeticVariable variable) {
            var coefficients = new float[variable.Context.Variables.Count];

            coefficients[variable.Index] = 1.0f;

            return new ArithmeticExpression(variable.Context, coefficients);
        }
    }
}

#pragma warning restore CS0660, CS0661

File: sdk/lib/constraint_arithmetic.cpp
#include "constraint_arithmetic.h"
#include "environment.h"

static void expand_coefficients(ArithmeticContext context, size_t *coefficient_count, float **coefficients, size_t manual_variable_count) {
    auto old_coefficient_count = *coefficient_count;

    *coefficient_count = context.variable_count;

    *coefficients = (float*)reallocate((void*)*coefficients, context.variable_count * sizeof(float));

    for(size_t i = old_coefficient_count; i < context.variable_count - manual_variable_count; i += 1) {
        (*coefficients)[i] = 0.0f;
    }
}

void add_arithmetic_constraint(
    ArithmeticExpression left,
    ArithmeticInequality inequality,
    ArithmeticExpression right
) {
    if(left.context != right.context) {
        return;
    }

    auto context = left.context;

    switch(inequality) {
        case ArithmeticInequality::Equal: break;

        case ArithmeticInequality::LessThanOrEqual:
        case ArithmeticInequality::GreaterThanOrEqual: {
            auto slack_variable = context->variable_count;
            context->variable_count += 1;

            context->is_variable_external = (bool*)reallocate((void*)context->is_variable_external, context->variable_count * sizeof(bool));

            context->is_variable_external[slack_variable] = false;

            expand_coefficients(*context, &left.coefficient_count, &left.coefficients, 1);

            float slack_coefficient;
            switch(inequality) {
                case ArithmeticInequality::LessThanOrEqual: {
                    slack_coefficient = 1.0f;
                } break;

                case ArithmeticInequality::GreaterThanOrEqual: {
                    slack_coefficient = -1.0f;
                } break;
            }

            left.coefficients[slack_variable] = slack_coefficient;
        } break;
    }

    auto constraint_index = context->constraint_count;

    context->constraint_count += 1;

    context->constraints = (ArithmeticContext::ArithmeticConstraint*)reallocate(
        (void*)context->constraints,
        context->constraint_count * sizeof(ArithmeticContext::ArithmeticConstraint)
    );
 
    context->constraints[constraint_index] = {
        left.constant,
        left.coefficient_count,
        left.coefficients,
        right.constant,
        right.coefficient_count,
        right.coefficients
    };
}

ArithmeticExpression operator+(ArithmeticExpression a, ArithmeticExpression b) {
    auto context = a.context;

    size_t coefficient_count;
    float *coefficients;
    if(a.coefficient_count > b.coefficient_count) {
        for(size_t i = 0; i < b.coefficient_count; i += 1) {
            a.coefficients[i] += b.coefficients[i];
        }

        deallocate(b.coefficients);

        coefficient_count = a.coefficient_count;
        coefficients = a.coefficients;
    } else {
        for(size_t i = 0; i < a.coefficient_count; i += 1) {
            b.coefficients[i] += a.coefficients[i];
        }

        deallocate(a.coefficients);

        coefficient_count = b.coefficient_count;
        coefficients = b.coefficients;
    }

    return {
        context,
        a.constant + b.constant,
        coefficient_count,
        coefficients
    };
}

ArithmeticExpression operator+(ArithmeticTerm a, ArithmeticTerm b) {
    auto context = a.context;

    auto coefficients = (float*)allocate(context->variable_count * sizeof(float));

    for(size_t i = 0; i < context->variable_count; i += 1) {
        coefficients[i] = 0.0f;
    }

    coefficients[a.variable_index] = a.coefficient;
    coefficients[b.variable_index] += b.coefficient;

    return {
        context,
        0.0f,
        context->variable_count,
        coefficients
    };
}

ArithmeticExpression operator+(ArithmeticTerm term, float constant) {
    auto coefficients = (float*)allocate(term.context->variable_count * sizeof(float));

    for(size_t i = 0; i < term.context->variable_count; i += 1) {
        coefficients[i] = 0.0f;
    }

    coefficients[term.variable_index] = term.coefficient;

    return {
        term.context,
        constant,
        term.context->variable_count,
        coefficients
    };
}

ArithmeticTerm::operator ArithmeticExpression() {
    auto coefficient_count = context->variable_count;
    auto coefficients = (float*)allocate(coefficient_count * sizeof(float));

    for(size_t i = 0; i < coefficient_count; i += 1) {
        coefficients[i] = 0.0f;
    }

    coefficients[variable_index] = coefficient;

    return {
        context,
        0.0f,
        coefficient_count,
        coefficients
    };
}

ArithmeticExpression operator+(ArithmeticExpression expression, ArithmeticTerm term) {
    auto context = expression.context;

    if(term.variable_index >= expression.coefficient_count) {
        auto old_coefficient_count = expression.coefficient_count;

        expression.coefficient_count = context->variable_count;

        expression.coefficients = (float*)reallocate((void*)expression.coefficients, context->variable_count * sizeof(float));

        for(size_t i = old_coefficient_count; i < context->variable_count; i += 1) {
            expression.coefficients[i] = 0.0f;
        }

        expression.coefficients[term.variable_index] = 0.0f;
    }

    expression.coefficients[term.variable_index] += term.coefficient;

    return expression;
}

ArithmeticVariable create_new_variable(ArithmeticContext *context) {
    auto index = context->variable_count;

    context->variable_count += 1;

    context->is_variable_external = (bool*)reallocate((void*)context->is_variable_external, context->variable_count * sizeof(bool));

    context->is_variable_external[index] = true;

    return {
        context,
        index
    };
}

ArithmeticExpression arithmetic_variable_to_expression(ArithmeticVariable variable) {
    auto coefficient_count = variable.context->variable_count;

    auto coefficients = (float*)allocate(coefficient_count * sizeof(float));

    for(size_t i = 0; i < coefficient_count; i += 1) {
        coefficients[i] = 0.0f;
    }

    coefficients[variable.index] = 1.0f;

    return {
        variable.context,
        0.0f,
        coefficient_count,
        coefficients
    };
}

bool solve_arithmetic_constraints(ArithmeticContext *context) {
    // Create full-sized tableau
    auto variable_count = context->variable_count;

    auto is_variable_external = context->is_variable_external;

    auto constraint_count = context->constraint_count;

    auto objective_constant = context->objective.constant;
    auto objective_coefficients = (float*)allocate(variable_count * sizeof(float));

    for(size_t i = 0; i < variable_count; i += 1) {
        objective_coefficients[i] = 0.0f;
    }

    for(size_t i = 0; i < context->objective.coefficient_count; i += 1) {
        objective_coefficients[i] = context->objective.coefficients[i];
    }

    deallocate(context->objective.coefficients);

    auto constraint_constants = (float*)allocate(constraint_count * sizeof(float));

    auto constraint_coefficients = (float*)allocate(constraint_count * variable_count * sizeof(float));

    for(size_t i = 0; i < constraint_count * variable_count; i += 1) {
        constraint_coefficients[i] = 0.0f;
    }

    for(size_t i = 0; i < constraint_count; i += 1) {
        auto constraint = context->constraints[i];

        constraint_constants[i] = constraint.left_constant;
        constraint_constants[i] -= constraint.right_constant;

        for(size_t j = 0; j < constraint.left_coefficient_count; j += 1) {
            constraint_coefficients[coefficient_index(i, j)] = constraint.left_coefficients[j];
        }

        deallocate(constraint.left_coefficients);

        for(size_t j = 0; j < constraint.right_coefficient_count; j += 1) {
            constraint_coefficients[coefficient_index(i, j)] -= constraint.right_coefficients[j];
        }

        deallocate(constraint.right_coefficients);
    }

    deallocate(context->constraints);

    auto constraint_variable_indices = (size_t*)allocate(constraint_count * sizeof(size_t));

    if(!solve(
        variable_count,
        constraint_count,
        is_variable_external,
        &objective_constant,
        objective_coefficients,
        constraint_variable_indices,
        constraint_constants,
        constraint_coefficients
    )) {
        deallocate(objective_coefficients);
        deallocate(constraint_coefficients);

        return false;
    }

    deallocate(objective_coefficients);
    deallocate(constraint_coefficients);

    context->solution_variable_indices = constraint_variable_indices;
    context->solution_constants = constraint_constants;

    return true;
}

float get_arithmetic_variable_value(ArithmeticVariable variable) {
    auto context = *variable.context;

    for(size_t i = 0; i < context.variable_count; i += 1) {
        if(context.solution_variable_indices[i] == variable.index) {
            return context.solution_constants[i];

            break;
        }
    }

    return 0.0f;
}

void arithmetic_minimize(ArithmeticExpression expression) {
    expression.context->objective = expression.context->objective + expression;
}

void arithmetic_maximize(ArithmeticExpression expression) {
    expression.context->objective = expression.context->objective - expression;
}